src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -686,16 +686,27 @@ HRESULT STDMETHODCALLTYPE CorProfiler::JITCompilationStarted(
 
       // insert the opcode and signature token as
       // additional arguments for the wrapper method
+      //
+      // IMPORTANT: Conditional branches may jump to this call instruction, so
+      // we cannot add the argument-loading instructions BEFORE this instruction,
+      // otherwise this will surface in an InvalidProgramException.
+      // Either begin loading the additional arguments starting with this ILInstr
+      // structure or make the current ILInstr a no-op and do all of the argument
+      // loading after this instruction.
       ILRewriterWrapper rewriter_wrapper(&rewriter);
       rewriter_wrapper.SetILPosition(pInstr);
-      rewriter_wrapper.LoadInt32(pInstr->m_opcode);
-      rewriter_wrapper.LoadInt32(method_def_md_token);
-      rewriter_wrapper.LoadInt64(reinterpret_cast<INT64>(module_version_id_ptr));
+      auto original_methodcall_opcode = pInstr->m_opcode;
+      pInstr->m_opcode = CEE_NOP;
 
+      // replace with a non-virtual call (CALL) to the instrumentation wrapper
       // always use CALL because the wrappers methods are all static
-      pInstr->m_opcode = CEE_CALL;
-      // replace with a call to the instrumentation wrapper
-      pInstr->m_Arg32 = wrapper_method_ref;
+      rewriter_wrapper.CallMemberAfter(wrapper_method_ref, false);
+      rewriter_wrapper.SetILPosition(pInstr->m_pNext);
```

Comment:
@zacharycmontoya Can you clarify the use of `CallMemberAfter` here? To me, it reads like:

1. current instruction is a NOP
2. calling wrapper after the current instruction
3. setting up the arguments last.

Shouldn't the args be set up first?

Reply:
That's correct, they do need to be set up first and that happens here, maybe in a not obvious way. The `ILInstr` structs are set up as linked lists, so this is what's happening in those lines of code:
1. We set the current instruction to NOP, followed by the call instruction. Result: `nop->call`.
1. We set the "current" instruction to the call instruction. Result `nop->call`
1. Each `LoadInt32` and `LoadInt64` insert BEFORE the current instruction. After the first LoadInt32 which loads the original opcode, the result is: `nop->loadint_opcode->call`
1. We keep inserting load instructions before the call instruction, so the final result is: `nop->loadint_opcode->loadint_mdtoken->loadint_mvid->call`

Reply:
ok got it. Thanks!

Reply:
I didn't see that the `ILRewriterWrapper` was putting `Load*` instructions in with `InsertBefore`.